frontend/public/components/secrets/create-secret.tsx

```diff
@@ -2,52 +2,77 @@
 import { WebHookSecretKey } from '../secret';
 
-export enum SecretTypes {
-  webhook = 'webhook',
+enum SecretTypeAbstraction {
   generic = 'generic',
+  source = 'source',
+  webhook = 'webhook',
+}
+
+export enum SecretType {
+  basicAuth = 'kubernetes.io/basic-auth',
+  sshAuth = 'kubernetes.io/ssh-auth',
+  opaque = 'Opaque'
 }
 
+const secretFormExplanation = {
+  source: 'Source secrets allow you to authenticate against the SCM server.',
+  webhook: 'Webhook secrets allow you to authenticate a webhook trigger.',
+};
+
+const determineDefaultSecretType = (typeAbstraction) => {
+  return typeAbstraction === 'source' ? SecretType.basicAuth : SecretType.opaque;
+};
+
 const determineSecretTypeAbstraction = (data) => {
-  return _.has(data, WebHookSecretKey) ? SecretTypes.webhook : SecretTypes.generic;
+  return _.has(data, WebHookSecretKey) ? SecretTypeAbstraction.webhook : SecretTypeAbstraction.source;
 };
 
-class BaseEditSecret_ extends SafetyFirst<BaseEditSecretProps_, BaseEditSecretState_> {
+const generateSecret = () => {
+  // http://stackoverflow.com/questions/105034/create-guid-uuid-in-javascript
+  const s4 = () => Math.floor((1 + Math.random()) * 0x10000).toString(16).substring(1);
+  return s4() + s4() + s4() + s4();
+};
+
+// withSecretForm returns SubformComponent which is a Higher Order Component for all the types of secret forms.
+const withSecretForm = (SubformComponent) => class SecretFormComponent extends React.Component<BaseEditSecretProps_, BaseEditSecretState_> {
   constructor (props) {
     super(props);
-    const existingObj = _.pick(props.obj, ['metadata', 'type']);
-    const existingData = _.get(props.obj, 'data');
-    const secret = _.defaultsDeep({}, props.fixed, existingObj, {
+    const inputObj = _.get(props.obj, 'data');
+    const existingSecret = _.pick(inputObj, ['metadata', 'type']);
+    const defaultSecretType = determineDefaultSecretType(this.props.secretTypeAbstraction);
+    const secret = _.defaultsDeep({}, props.fixed, existingSecret, {
       apiVersion: 'v1',
       data: {},
       kind: 'Secret',
       metadata: {
         name: '',
       },
-      type: 'Opaque',
+      type: defaultSecretType,
     });
 
     this.state = {
-      secretType: this.props.secretType || determineSecretTypeAbstraction(existingData),
+      secretTypeAbstraction: this.props.secretTypeAbstraction,
       secret: secret,
       inProgress: false,
-      type: secret.type,
-      stringData: _.mapValues(existingData, window.atob),
+      type: defaultSecretType,
+      stringData: _.mapValues(_.get(inputObj, 'data'), window.atob),
     };
     this.onDataChanged = this.onDataChanged.bind(this);
     this.onNameChanged = this.onNameChanged.bind(this);
     this.save = this.save.bind(this);
   }
   onDataChanged (secretsData) {
-    this.setState({stringData: {...secretsData}});
+    this.setState({
+      stringData: {...secretsData.stringData},
+      type: secretsData.authenticationType,
```

Comment:
It sort of breaks the abstraction if the SecretFormComponent knows about authenticationType. Can this be changed to just "type"?

Reply:
makes sense